rfc/java11.md

```diff
@@ -0,0 +1,164 @@
+| **Author(s)**  | Michael Plump ([`@plumpy`](https://github.com/plumpy)) |
+| **SIG / WG**   | Platform SIG |
+
+## Overview
+
+To keep current with the evolution of Java, Spinnaker needs to move from Java 8
+to Java 11. This RFC documents a process and timeline for doing so.
+
+### Goals and Non-Goals
+
+Goals:
+
+*   Provide a process for moving official Spinnaker builds to the Java 11 JRE
+*   Discuss ways we will encourage people using non-offical builds to also move
+    to the Java 11 JRE
+*   Provide a timeline for the final switch to a Java 11 JDK along with criteria
+    to decide if the timeline should be delayed
+
+Non-Goals:
+
+*   Providing a rationale for upgrading Java; all active Java projects need to
+    update their JDK version eventually as libraries we depend upon will also
+    depend on new versions of the JDK
+*   Debating which Java version we should upgrade to; we take it as a given that
+    only LTS releases are acceptable for most organizations, especially for
+    running critical infrastructure like Spinnaker
+*   Documenting and solving technical issues that present themselves in the
+    migration; those will be handled separately, but this document concerns
+    itself primarily with the larger processes and timeline
+*   Detailing the required release system changes; since the releases are still
+    entirely maintained by Google, these changes aren't particularly relevant to
+    the wider community
+
+### JRE vs. JDK
+
+There are two components of Java of interest to this RFC. The Java Runtime
+Environment (JRE) is the `java` binary and associated libraries used to _run_
+Spinnaker services. The Java Development Kit (JDK) is the `javac` binary and
+associated libraries used to _compile_ Spinnaker.
+
+Once we upgrade to the Java 11 JDK, Spinnaker will no longer be able to run in a
+Java 8 JRE. So the first (more complicated) step is to get everyone migrated to
+a Java 11 JRE. Only after that is done can we switch to the Java 11 JDK.
+
+The Java 11 JRE has some fairly large changes, most notably the deprecation of
+the long-standing CMS garbage collector. Large organizations with highly-tuned
+JVMs have shown hesitancy to upgrade until the new runtime has undergone
+extensive testing. (Both Google and Netflix are in this category.)
+
+For official Spinnaker containers, we control the JRE used to run Spinnaker. For
+those releases, this proposal takes a relatively slow ramp to the Java 11 JRE
+and provides an escape hatch back to Java 8 for users who experience problems.
+
+There are also some unknown number of customers who aren't using the Spinnaker-
+provided containers. We will need to do some outreach to make sure these users
+convert to a Java 11 JRE too.
+
+## Timeline
+
+### 1.18 (early January)
+
+For our container builds, we'll first move just two (of approximately ten)
+services to the new JRE. Thee two services will be `igor` and `orca`. By
```

Comment:
Out of curiosity, how were these services chosen?

Reply:
Somewhat arbitrarily. My criteria:
1) I wanted 2–3 services to start
2) I wanted some small and some a little "meatier"
3) but since the GC is the main thing I'm worried about (mostly because that's what I've heard Google is most concerned about), I wanted to avoid `clouddriver` initially, which I imagine will be the most sensitive to GC changes

Happy to discuss picking other services if you think there are better candidates...

Reply:
I don't have any specific suggestions, I was just curious. +1 to having small/large. I'm a _touch_ worried about Orca because of how critical it is but I trust that the Platform SIG (I assume this falls under their realm) has thought through that.

Reply:
I don't have any specific suggestions, I was just curious. +1 to having small/large. I'm a _touch_ worried about Orca because of how critical it is but I trust that the Platform SIG (I assume this falls under their realm) has thought through that.

Reply:
The platform SIG hasn't discussed this yet, but I'm going to bring it up at our next meeting :grin:

Reply:
I'm not particularly concerned with selecting Orca, but if we want a "larger" service that iterates with less cadence, Front50 may be a good choice.

Reply:
Switched `orca` to `front50`.